File: lib/awj.js
// Minimal AWJ client for Analog Way LivePremier / Alta 4K / Midra 4K.
//
// Wire format: TCP 10606, one JSON object per message, terminated by ASCII 0x04
// (not a newline). Exactly one `op`, only ever `get` or `replace`.
//
// This client is deliberately GET-ONLY. Every property this tool reads is
// declared readOnly in the device's own model, so there is nothing here that
// can change device state — see AGENTS.md.

import net from 'node:net';

const EOT = 0x04;
const EOT_STR = '\x04';

export class AwjError extends Error {
  constructor(code, message, path) {
    super(message || `AWJ error ${code}`);
    this.name = 'AwjError';
    this.code = code;
    this.path = path;
  }
}

export class AwjClient {
  /**
   * @param {object} opts
   * @param {string} opts.host
   * @param {number} [opts.port=10606]
   * @param {number} [opts.timeout=5000]  per-request timeout, ms
   */
  constructor({ host, port = 10606, timeout = 5000 }) {
    this.host = host;
    this.port = port;
    this.timeout = timeout;
    this.socket = null;
    this.buf = Buffer.alloc(0);
    this.queue = [];
    this.closed = false;
  }

  connect() {
    return new Promise((resolve, reject) => {
      const sock = net.createConnection({ host: this.host, port: this.port });
      sock.setNoDelay(true);

      const onFail = (err) => {
        sock.destroy();
        reject(err);
      };
      const connectTimer = setTimeout(
        () => onFail(new Error(`timed out connecting to ${this.host}:${this.port}`)),
        this.timeout,
      );

      sock.once('connect', () => {
        clearTimeout(connectTimer);
        sock.removeListener('error', onFail);
        this.socket = sock;

        sock.on('data', (chunk) => this._onData(chunk));
        sock.on('error', (err) => this._failAll(err));
        sock.on('close', () => {
          this.closed = true;
          this._failAll(new Error('connection closed by device'));
        });

        resolve(this);
      });
      sock.once('error', onFail);
    });
  }

  _onData(chunk) {
    this.buf = Buffer.concat([this.buf, chunk]);
    let idx;
    while ((idx = this.buf.indexOf(EOT)) !== -1) {
      const raw = this.buf.subarray(0, idx).toString('utf8');
      this.buf = this.buf.subarray(idx + 1);
      const waiter = this.queue.shift();
      if (!waiter) continue; // unsolicited push; this tool does not subscribe
      clearTimeout(waiter.timer);
      let msg;
      try {
        msg = JSON.parse(raw);
      } catch {
        waiter.reject(new Error(`unparseable AWJ frame: ${raw.slice(0, 120)}`));
        continue;
      }
      if (msg.error) {
        waiter.reject(new AwjError(msg.error.code, msg.error.message, waiter.path));
      } else {
        waiter.resolve(msg.value);
      }
    }
  }

  _failAll(err) {
    const q = this.queue;
    this.queue = [];
    for (const w of q) {
      clearTimeout(w.timer);
      w.reject(err);
    }
  }

  /**
   * Read one leaf. Containers legitimately resolve to {} — AWJ cannot enumerate.
   * @param {string} path
   */
  get(path) {
    if (!this.socket || this.closed) return Promise.reject(new Error('not connected'));
    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        const i = this.queue.findIndex((w) => w.timer === timer);
        if (i !== -1) this.queue.splice(i, 1);
        reject(new Error(`timed out reading ${path}`));
      }, this.timeout);

      this.queue.push({ resolve, reject, timer, path });
      this.socket.write(JSON.stringify({ op: 'get', path }) + EOT_STR);
    });
  }

  /** get() that resolves to `undefined` instead of throwing on E12 (path absent). */
  async tryGet(path) {
    try {
      return await this.get(path);
    } catch (err) {
      if (err instanceof AwjError && err.code === 'E12') return undefined;
      throw err;
    }
  }

  close() {
    this.closed = true;
    if (this.socket) this.socket.destroy();
    this.socket = null;
  }
}

File: src-tauri/src/awj.rs
//! Minimal AWJ client — the desktop app's equivalent of `lib/awj.js`.
//!
//! AWJ: TCP 10606, one JSON object per message, terminated by the byte 0x04.
//! Exactly one `op`, and this client only ever sends `get`.
//!
//! GET-ONLY BY CONSTRUCTION. There is one function that writes to the socket
//! and it builds `{"op":"get",…}` from a literal. Nothing here can emit a
//! write, and nothing should be added that can — see AGENTS.md.

use std::io::{Read, Write};
use std::net::{TcpStream, ToSocketAddrs};
use std::time::Duration;

const EOT: u8 = 0x04;

pub struct Awj {
    stream: TcpStream,
    buf: Vec<u8>,
}

#[derive(Debug)]
pub enum AwjError {
    Connect(String),
    Io(String),
    /// The device answered, but with an error code — `E12` means "no such path".
    Device(String),
}

impl std::fmt::Display for AwjError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            AwjError::Connect(m) => write!(f, "{m}"),
            AwjError::Io(m) => write!(f, "{m}"),
            AwjError::Device(c) => write!(f, "device error {c}"),
        }
    }
}

impl Awj {
    pub fn connect(host: &str, port: u16, timeout: Duration) -> Result<Self, AwjError> {
        let addr = (host, port)
            .to_socket_addrs()
            .map_err(|e| AwjError::Connect(format!("cannot resolve {host}: {e}")))?
            .next()
            .ok_or_else(|| AwjError::Connect(format!("cannot resolve {host}")))?;

        let stream = TcpStream::connect_timeout(&addr, timeout)
            .map_err(|e| AwjError::Connect(format!("cannot reach {host}:{port} — {e}")))?;
        stream.set_read_timeout(Some(timeout)).ok();
        stream.set_write_timeout(Some(timeout)).ok();
        stream.set_nodelay(true).ok();

        Ok(Self { stream, buf: Vec::new() })
    }

    /// Send one `get` and read its reply. The only write on this socket.
    pub fn get(&mut self, path: &str) -> Result<serde_json::Value, AwjError> {
        let req = serde_json::json!({ "op": "get", "path": path });
        let mut bytes = serde_json::to_vec(&req).map_err(|e| AwjError::Io(e.to_string()))?;
        bytes.push(EOT);
        self.stream
            .write_all(&bytes)
            .map_err(|e| AwjError::Io(format!("write failed: {e}")))?;

        let frame = self.read_frame()?;
        let msg: serde_json::Value =
            serde_json::from_slice(&frame).map_err(|e| AwjError::Io(format!("bad frame: {e}")))?;

        if let Some(err) = msg.get("error") {
            let code = err
                .get("code")
                .and_then(|c| c.as_str())
                .unwrap_or("UNKNOWN")
                .to_string();
            return Err(AwjError::Device(code));
        }
        Ok(msg.get("value").cloned().unwrap_or(serde_json::Value::Null))
    }

    /// `get` that yields `None` for a path the device does not have (E12),
    /// mirroring `tryGet` on the JavaScript side.
    pub fn try_get(&mut self, path: &str) -> Option<serde_json::Value> {
        match self.get(path) {
            Ok(v) => Some(v),
            Err(_) => None,
        }
    }

    fn read_frame(&mut self) -> Result<Vec<u8>, AwjError> {
        loop {
            if let Some(i) = self.buf.iter().position(|&b| b == EOT) {
                let frame: Vec<u8> = self.buf.drain(..=i).collect();
                return Ok(frame[..frame.len() - 1].to_vec());
            }
            let mut chunk = [0u8; 8192];
            let n = self
                .stream
                .read(&mut chunk)
                .map_err(|e| AwjError::Io(format!("read failed: {e}")))?;
            if n == 0 {
                return Err(AwjError::Io("connection closed by device".into()));
            }
            self.buf.extend_from_slice(&chunk[..n]);
        }
    }
}
